fix Misstep

ameliatqy opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->

Fixes #2455. This used to be https://github.com/PyTorchLightning/pytorch-lightning/pull/2475 but I messed up on rebasing and to be safe, created a new pull request. This is a Draft PR.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

<!-- For CHANGELOG separate each item in unreleased section by a blank line to reduce collisions -->

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Yes, I am new to contributing to open-source and actually find it quite fun. May be time to consider a new hobby...

awaelchli commented on pytorch_lightning/trainer/training_loop.py:
```diff
@@ -504,6 +505,9 @@ def run_training_epoch(self):
         self.run_on_epoch_end_hook(model)
 
+        # increate global step by one to progress to the next epoch
+        self.global_step += 1
```
I think you also need to do do self.increment_accumulated_grad_global_step() here, otherwise the total_train_batch_idx does not get updated. I think that's why the tests are failing.

ameliatqy replied:
I made the change and it works as expected. So I have changed it. 

However, the tests are still failing. Actually, I think I am confused about the test. I ran the local test on my pull request and got this total summary result at the end:
<img width="786" alt="Screen Shot 2020-07-03 at 12 40 46 AM" src="https://user-images.githubusercontent.com/33330101/86444629-e254aa00-bcc5-11ea-9c69-253b481b49f2.png">

I also ran the test over the original Pytorch Lightning master branch and also got this total summary result at the end:
<img width="756" alt="Screen Shot 2020-07-03 at 12 34 03 AM" src="https://user-images.githubusercontent.com/33330101/86444021-fcda5380-bcc4-11ea-8588-a09edb94f565.png">

Am I making a mistake somewhere? Should I have based my branch off on something other than the master branch?

awaelchli commented on pytorch_lightning/loggers/base.py:
```diff
@@ -102,7 +102,24 @@ def _reduce_agg_metrics(self):
             agg_mets = self._metrics_to_agg[0]
         else:
-            agg_mets = merge_dicts(self._metrics_to_agg, self._agg_key_funcs, self._agg_default_func)
+            # check if dictionary keys are unique
+            agg_keys = set([key for met in self._metrics_to_agg for key in met.keys()])
+            num_keys = sum([len(met) for met in self._metrics_to_agg])
+
+            # exclude 'epoch' because it is a metric automatically added in by log_metrics and will count as a
+            # duplicate. If you want to get rid of this, I would suggest you should get rid of `scalar_metrics[
```
are the changes here really related to the global step issue?
If possible could we make a separate PR for this? we only want to fix one thing at a time.

ameliatqy replied:
I can see it being a separate issue. But, this piece of code is necessary for the code to run smoothly. Otherwise, when we fix the step issue (i.e. combining the last training batch + training_epoch_end + validation_epoch_end metrics in the same step), there will be an error combining the three pieces of data together because the original `merge_dict()` in `_reduce_agg_metrics()` requires that all the metric dictionaries have the same keys - which for these three types of data is usually not the case.

I don't mind doing a separate pull request for this though - but it will probably result in most tests failing for this pull request, which we will probably have to make sure to fix in the new pull request.

Borda commented:
@ameliatqy I have rebased it to master, can we finish it in this release?

pep8speaks commented:
Hello @ameliatqy! Thanks for updating this PR. 







* In the file [`pytorch_lightning/trainer/training_loop.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/54db53f6b8203a0ab5a08dfaa1b33461ab93a644/pytorch_lightning/trainer/training_loop.py):

> [Line 537:25](https://github.com/PyTorchLightning/pytorch-lightning/blob/54db53f6b8203a0ab5a08dfaa1b33461ab93a644/pytorch_lightning/trainer/training_loop.py#L537): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

ameliatqy commented:
> @ameliatqy I have rebased it to master, can we finish it in this release?

I have worked on the pull request and the checks have all passed. How do I get it approved? I  see that I need three approving reviews - am I responsible for finding people to review it or will the reviewers be assigned by someone else?

awaelchli commented:
> > @ameliatqy I have rebased it to master, can we finish it in this release?
> 
> I have worked on the pull request and the checks have all passed. How do I get it approved? I see that I need three approving reviews - am I responsible for finding people to review it or will the reviewers be assigned by someone else?

Sorry for this late response.
Anyone in the community can review, but only the reviews of the contributors count towards a merge decision. 

I have two questions:
1. Does the issue with global step still exist on master?
2. Why do we need a decrement in the global_step? I would prefer a solution where this is not needed. Also, there was recently a fix #2853 in accumulate batches for the last batch, so I have the feeling this relates here.

ameliatqy commented:
> > > @ameliatqy I have rebased it to master, can we finish it in this release?
> > 
> > 
> > I have worked on the pull request and the checks have all passed. How do I get it approved? I see that I need three approving reviews - am I responsible for finding people to review it or will the reviewers be assigned by someone else?
> 
> Sorry for this late response.
> Anyone in the community can review, but only the reviews of the core team count towards a merge decision.
> 
> I have two questions:
> 
> 1. Does the issue with global step still exist on master?
> 2. Why do we need a decrement in the global_step? I would prefer a solution where this is not needed. Also, there was recently a fix #2853 in accumulate batches for the last batch, so I have the feeling this relates here.

1. Yes, the issue still exists on master - I ran a test today and I still get the same error with the current master branch
2. I decremented it because, as discussed in issue #2455 , I plan to put the results of training_epoch_end in the same step as the results of the last batch when logging metrics. As such, the steps increment in the batch loop as usual until we get to the last step, in which we will increment the global_step at the very end of the function. The issue here is that we want to break the loop when `batch_idx >= self.num_training_batches`, even on the first loop. On the first loop, you can break the loop with no repercussions. But there are multiple loops, we would've incremented the global_step/total_batch_idx by one extra step - so the results of the last batch and training_epoch_end aren't stored together anymore. So I had to decrement it to the global_step/total_batch_idx of the last loop so that when` self.run_training_epoch_end()` is executed, the results of the last batch and training_epoch_end are stored together and the number of runs will sync together (as you described in https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653004576)

However, I understand you and I got rid of the decrement function - you can look at my new solution and see what you think. Additionally, the pull request that you linked for fixing accumulate batches is a different fix. That fixes a mechanism in which an action for the last batch wasn't executed when it was supposed to. This isn't the case here, here we can fixing a case in which a dictionary may have different keys (i.e. keys from the last batch training and keys from training_epoch_end). You can review my reasoning here which provides more details: https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653256670

ameliatqy commented:
Just so you know, I had to remove ` self.logger.save()` from the function `log_metrics` because it was resetting `self._metrics_to_agg` every time log_metrics was called. This was not letting metrics from the same self.global_step aggregate and consequently making my epochs/step 'out of step' as talked about in https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653004576

After fixing that issues, I realize now that some tests aren't passing. Here is why:
```
FAILED tests/loggers/test_all.py::test_loggers_fit_test[TensorBoardLogger] - ...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[CometLogger] - Assert...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[MLFlowLogger] - Asser...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[NeptuneLogger] - Asse...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[TestTubeLogger] - Ass...
FAILED tests/loggers/test_all.py::test_loggers_fit_test[WandbLogger] - Assert...
```
These tests aren't passing because they expect the step of the test to be different from the validation and training. In order to keep the epoch and steps in lines (i.e. number of total steps = no batches*number of epochs), I believe the the test step should be the same as the training_epoch_end, validation_epoch_end and the results from the last batch.

```
FAILED tests/loggers/test_base.py::test_adding_step_key - assert 1 == 2
```
This test fails because it expects  validation_epoch_end and training_epoch_end to share the same step number. As discussed, this is not possible if we want to  keep the epoch and steps in lines (i.e. number of total steps = no batches*number of epochs).

So I guess I have a design question. Do you still want the epochs and step numbers to line up (i.e. number of total steps = no batches*number of epochs like in https://github.com/PyTorchLightning/pytorch-lightning/issues/2455#issuecomment-653004576)? Or should the training_epoch_end/validation_epoch_end/test_epoch_end have their own step. Please note that this will result in an offset as described in the comment linked (which I believe is undesired but let me know).

williamFalcon commented:
please rebase master since this code is super old haha

ameliatqy commented:
> please rebase master since this code is super old haha

Interesting, I thought I rebased my branch today XD I just rebased it again - does it look right? Rebasing is obviously not my strong point XD

Borda commented:
@ameliatqy how about this one, mind rebase and resolve conflicts? 🐰

ameliatqy commented:
> @ameliatqy how about this one, mind rebase and resolve conflicts? 🐰

Sure, I'd be happy to. But to fully resolve this pull request, could someone answer my question in https://github.com/PyTorchLightning/pytorch-lightning/pull/2478#issuecomment-675183459 ? I can fully finish up the pull request without an answer to this question.

EspenHa commented on pytorch_lightning/trainer/logging.py:
```diff
@@ -73,8 +74,6 @@ def log_metrics(self, metrics, grad_norm_dic, step=None):
         # log actual metrics
         if self.is_global_zero and self.logger is not None:
             self.logger.agg_and_log_metrics(scalar_metrics, step=step)
-            self.logger.save()
```
How is this related to https://github.com/PyTorchLightning/pytorch-lightning/issues/3398#issuecomment-691011447?
@SkafteNicki Maybe you know something about this also?

awaelchli commented:
This is now fixed on master. There is no offset anymore and training epoch end logs at the correct step.
